File: 3133-minimum-array-end/3133-minimum-array-end.c
long long minEnd(int n, int x) {
    n--;
    int i = 0;
    long long res = 0;
    while(n||x) {
        if (x & 1) {
            res |= (long long)1 << i;
        } else {
            res |= (long long)(n&1) << i;
            n >>= 1;
        }
        i++;
        x >>= 1;
    }
    return res;
}

File: 3133-minimum-array-end/3133-minimum-array-end.py
class Solution:
    def minEnd(self, n: int, x: int) -> int:
        if n == 1:
            return x
        res = 0
        i = 0
        n -= 1
        while n or x:
            if x&1:
                res |= 1 << i
            else:
                res |= (n&1) << i
                n >>= 1
            x >>= 1
            i += 1
        return res